src/core/prototypes/state.ts

```diff
@@ -72,6 +85,209 @@ export class ChartStateManager {
     }
   }
 
+  public resetDifference() {
+    this.chartOrigin = deepClone(this.chart);
+  }
+
+  // eslint-disable-next-line max-lines-per-function
+  public hasUnsavedChanges() {
+    const origin = this.chartOrigin;
+    const chart = this.chart;
+
+    const currentProperties = chart.properties;
+    const originProperties = origin.properties;
+
+    try {
+      expect_deep_approximately_equals(currentProperties, originProperties, 0);
+    } catch {
+      return true;
+    }
+
+    if (origin.constraints.length !== chart.constraints.length) {
+      return true;
+    } else {
+      for (let index = 0; index < origin.constraints.length; index++) {
+        const originConstringts = origin.constraints[index];
+        const current = chart.constraints[index];
+        if (originConstringts.type != current.type) {
+          return true;
+        }
+        if (
+          originConstringts.attributes.element != current.attributes.element
+        ) {
+          return true;
+        }
+        if (originConstringts.attributes.gap != current.attributes.gap) {
+          return true;
+        }
+        if (
+          originConstringts.attributes.attribute != current.attributes.attribute
+        ) {
+          return true;
+        }
+        if (
+          originConstringts.attributes.targetAttribute !=
+          current.attributes.targetAttribute
+        ) {
+          return true;
+        }
+        if (
+          originConstringts.attributes.targetElement !=
+          current.attributes.targetElement
+        ) {
+          return true;
+        }
+      }
+    }
+
+    const chartElements = [...forEachObject(chart)];
+    const originElements = [...forEachObject(origin)];
+
+    if (chartElements.length != originElements.length) {
```

Comment:
This section seems dangerous to me because when the schema changes, it is not obvious that this will also need a check added

Reply:
Check for what? If schema changes, generator will gives new items

Reply:
For example, when a new property is added to the schema. Code below is looking at `.kind`, `.glyph`, `.mark` etc.

Reply:
If schema changes you need to update generator too: https://github.com/microsoft/charticulator/blob/53030a453db62d4b095b789e7c3790c76bb269a1/src/core/prototypes/common.ts#L455-L475